prow/crier/reporters/criercommonlib/shardedlock.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package criercommonlib contains shared lib used by reporters
+package criercommonlib
+
+import (
+	"context"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/semaphore"
+)
+
+// SimplePull contains info for identifying a shard
+type SimplePull struct {
+	org, repo string
+	number    int
+}
+
+// NewSimplePull creates SimplePull
+func NewSimplePull(org, repo string, number int) *SimplePull {
+	return &SimplePull{org: org, repo: repo, number: number}
+}
+
+// ShardedLock contains sharding information based on PRs
+type ShardedLock struct {
+	mapLock *semaphore.Weighted
+	locks   map[SimplePull]*semaphore.Weighted
```

Comment:
Is there a benefit to using semaphores rather than mutexes?

Reply:
this code is from github reporter, simply copy/pasted for consistency reason. I'm not quite sure why using semaphore here, maybe @alvaroaleman still remembers?

Reply:
The semaphore usage was added in 3b97e7abebb406aeb487a3080ebebdf758b989da because its `Acquire` will respect the timeout of a context, there is nothing similiar available for the standard sync.Mutex, to my knowledge.

Reply:
I've stole your comment and left it in the code for future reference